### ISSUE TITLE
Single converted from Observable adapted by RxJavaCallAdapterFactory does not obey Single contract

### ISSUE THREAD

**Issue by tazle** [square/retrofit#3414]
It appears that when an Observable produced using RxJavaCallAdapterFactory gets adapted to a Single with `toSingle()`, the resulting Single doesn't obey Single's contract of throwing *OnErrorNotImplementedException* if the Single emits an error and it has been subscribed-to with `subscribe(x -> doSomething(x))`, i.e. there is no onError handler.

The cause seems to be https://github.com/square/retrofit/commit/308fe19a29d6d69fb4a9e99d3de434163cc95c7a which passes the error to RxJava's error handler, which by default swallows it.

I'll do some work on a test case.

**Comment by tazle** [comment]
Test case added.

**Comment by tazle** [comment]
Apparently the native Single conversion also suffers from the same issue. Added test for that too to the same branch.

**Comment by JakeWharton** [comment]
This looks like it's working as expected. `subscribe()` is not a blocking operation.

**Comment by tazle** [comment]
Isn't it, by default?

**Comment by JakeWharton** [comment]
For RxJava 1 and RxJava 2 the factory `create()` method will execute blocking HTTP requests on the subscribing thread, but that doesn't mean `subscribe` will propagate exceptions because that synchronous behavior is easily defeated with a `subscribeOn(io())`. We have no control over the behavior of RxJava's operators which are downstream of the returned `Single` and so we can only deliver the error to the error callback or, when downstream is unsubscribed, the plugin.

**Comment by tazle** [comment]
So you are saying that because it could be async, there is no point in propagating the exception in the sync case?

**Comment by JakeWharton** [comment]
Yes, but this is RxJava's behavior. We have no control over it.

**Comment by tazle** [comment]
Yes you do. The code introduced in 308fe19 is what swallows the exception in question by passing it to the RxJava error handler plugin, which by default does nothing.

It appears that the change was originally done to avoid swallowing the errors in some other (async) case.

**Comment by JakeWharton** [comment]
Right but the original code intended the exception to reach the handler it was only because of unexpected behavior of `throwIfFatal` that it did not. The choices are either to bubble up to the carrier thread (which is almost always a thread pool whose uncaught exception handler is non-existent and thus does nothing by default) or it's the plugin handler. The difference is that the latter is used by the rest of the library to deliver errors which cannot be sent through the `onError` pathway for one reason or another.

**Comment by tazle** [comment]
Ok, so it sounds like we should set up the the handler plugin.

We ran into this when replacing some Retrofit calls with simple `Single.error(new RuntimeException())` mocks suddenly exposed subscribe() calls without error handling in our code base.
